### PR TITLE
perf(core): async op pseudo-codegen and performance work

### PR DESCRIPTION
**Pull request by mmastrac**
Performance:

```
async_ops.js: 760k -> 1030k (!)
async_ops_deferred.js: 730k -> 770k
Deno.serve bench: 118k -> 124k
WS test w/ third_party/prebuilt/mac/load_test 100 localhost 8000 0 0: unchanged

Startup time: approx 0.5ms slower (13.7 -> 14.2ms)
```

### CODE REVIEW

**Comment by petamoriken** [comment]
We cannot use`eval` and `new Function` in core, because it denies `--disallow-code-generation-from-strings`V8 flag.
FYI: #17899

File: cli/bench/async_ops_deferred.js [hunk 1]
```diff
@@ -17,4 +17,5 @@ async function bench(fun) {
 }
 
 const core = Deno[Deno.internal].core;
-bench(() => core.opAsync("op_void_async_deferred"));
+const ops = core.ops;
+bench(() => ops.op_void_async_deferred());
```
Comment (bartlomieju) [inline cli/bench/async_ops_deferred.js:21]:
```suggestion
const op_void_async_deferred = core.ops.op_void_async_deferred;
bench(() => op_void_async_deferred());
```

File: core/01_core.js [hunk 2]
```diff
@@ -221,59 +257,346 @@
...
+          `Too many arguments for async op codegen (length of ${opName} was ${
+            originalOp.length - 1
+          })`,
+        );
+    }
+    return (ops[opName] = fn);
   }
 
   function opAsync2(name, arg0, arg1) {
     const id = nextPromiseId++;
```
Comment (mmastrac) [inline core/01_core.js:556]: If we could get rid of the requirement that promises resolving immediately have no promise ID, we can skip the `movePromise` step. Unfortunately there's a lot of code requiring promise IDs and I'm not sure we can do this yet.

File: core/runtime.rs [hunk 3]
```diff
@@ -3737,21 +3737,6 @@ assertEquals(1, notify_return_value);
...
   }
 
-  #[test]
-  fn test_core_js_stack_frame() {
-    let mut runtime = JsRuntime::new(RuntimeOptions::default());
-    // Call non-existent op so we get error from `core.js`
-    let error = runtime
-      .execute_script_static(
-        "core_js_stack_frame.js",
-        "Deno.core.opAsync('non_existent');",
```
Comment (mmastrac) [inline core/runtime.rs:3747]: These are async ops and the error for non-existent ops was made async as well. opAsync and opAsync2 should be going away soon, so all of these tests will eventually go away too.

File: core/01_core.js [hunk 4]
```diff
@@ -221,59 +257,359 @@
...
+        };
+        break;
+  `;
+}
+  */
+
+  // This function is called once per async stub
+  function asyncStub(opName, args) {
+    setUpAsyncStub(opName);
+    return ops[opName].apply(undefined, args);
```
Comment (petamoriken) [inline core/01_core.js:308]: Please use primordials

```suggestion
    return ReflectApply(ops[opName], undefined, args);
```

**Review by petamoriken — CHANGES_REQUESTED** [review]

**Review by littledivy — COMMENTED** [review]
Nice PR! 

I don't like the repetition with `core.ensureFastOps`. 

We've had `initializeAsyncOps` in the past that would be called during bootstrap which overrides `ops[opName]` for all async ops to setup an async stub.

See https://github.com/denoland/deno/pull/16428 (Similar to this PR but uses `eval()`)

**Comment by mmastrac** [comment]
@littledivy Thanks! 😄  I don't like the repetition either, but I guess `eval` is out which makes this a much more challenging problem. I'd like to work on some way to do codegen for these ops that allows us to avoid that.

The big challenge is keeping all of these functions in one file since they're all working with the same promise ringbuffer that we want the closures to capture directly.. 

For now this works, but I'd like to brainstorm having some sort of opt-in codegen for extensions (maybe handlebars-based?).

File: core/01_core.js (hunk 2 above)
Comment (bartlomieju) [reply core/01_core.js:568]: I think we can pass promiseId to Rust but create the actual promise only after dispatching we didn't receive immediate response. WDYT?

Comment (bartlomieju) [reply core/01_core.js:568]: I see that you already did that. I think this is already an improvement that you are lazily creating promises only when they are needed and passing numbers shouldn't be that bad in terms of performance.

**Comment by bartlomieju** [comment]
I did check all async ops we have and I found we have 2 ops with arity 5 and 3 ops with arity 7. I think we should probably limit ourselves to some reasonable number like max no of args equal to 4 or 5. All the ops that have higher arity should be rewritten to pass these args in a tuple.

One more suggestion I have is to change the name of `Deno.core.ensureFastOps` to something like `Deno.core.generateAsyncOpHandler`. `ensureFastOps` might be misleading, because these ops are not guaranteed to use V8 Fast Call API. 

Other than that I think we should land this PR ASAP and keep iterating on it since it provides a huge speed up.

**Review by bartlomieju — APPROVED** [review]
LGTM, discussed offline. We are trading a little bit of startup time for higher throughput on purpose. We will circle back and fix this regression after migrating all ops to the new system.

**Pull request merged**